src/ensemble/EnsembleBase.cpp

```diff
@@ -58,29 +57,40 @@ void Ensemble::readXML(XMLfileUnits& xmlconfig) {
 
 		xmlconfig.getNodeValue("@type", mixingruletype);
 		Log::global_log->info() << "Mixing rule type: " << mixingruletype << std::endl;
-		if("LB" == mixingruletype) {
+		if ("LB" == mixingruletype) {
 			mixingrule = new LorentzBerthelotMixingRule();
 
 		} else {
 			Log::global_log->error() << "Unknown mixing rule " << mixingruletype << std::endl;
 			Simulation::exit(1);
 		}
 		mixingrule->readXML(xmlconfig);
-		_mixingrules.push_back(mixingrule);
 
-		/*
-		 * Mixing coefficients
-		 *
-		 * TODO: information of mixing rules (eta, xi) is stored in Domain class and its actually in use
-		 * --> we need to decide where this information should be stored in future, in ensemble class,
-		 * in the way it is done above?
-		 *
-		 */
-		double xi, eta;
-		xmlconfig.getNodeValue("xi", xi);
-		xmlconfig.getNodeValue("eta", eta);
-		dmixcoeff.push_back(xi);
-		dmixcoeff.push_back(eta);
+		const int cid1 = mixingrule->getCid1();
+		const int cid2 = mixingrule->getCid2();
+		// Check if cid is larger than number of components
+		// cid starts with 0 and cid2 is always larger than cid1
+		if (cid2 >= numComponents) {
+			Log::global_log->error() << "Mixing: cid=" << cid2+1 << " is larger than number of components ("
+									 << numComponents << ")" << std::endl;
+			Simulation::exit(1);
+		}
+		_mixingrules[cid1][cid2] = mixingrule;
+	}
+	// Use xi=eta=1.0 as default if no rule was specified
+	for (int cidi = 0; cidi < numComponents; ++cidi) {
+		for (int cidj = cidi+1; cidj < numComponents; ++cidj) {  // cidj is always larger than cidi
+			if (_mixingrules[cidi].count(cidj) == 0) {
+				// Only LorentzBerthelot is supported until now
+				LorentzBerthelotMixingRule* mixingrule = new LorentzBerthelotMixingRule();
```

Comment:
I'd MUCH more prefer if we could use smartpointers here, e.g. `unique_ptr`.

Reply:
I have used `shared_ptr` now. `unique_ptr` does not work due to `EnsembleBase::setMixingrule` (I guess).

Reply:
good enough for now :)